retrieving/retrieving-and-querying-mongodb-documents-with-go.md

```diff
@@ -47,108 +47,117 @@ Knowing the fields that exist in our data will be important to us when it comes
+```go
 cursor, err := episodesCollection.Find(ctx, bson.M{})
 if err != nil {
     log.Fatal(err)
 }
-for cursor.Next(ctx) {
-    var episode bson.D
-    cursor.Decode(&episode)
-    fmt.Println(episode)
-}
-cursor.Close(ctx)
+var episodes []bson.M
+cursor.All(ctx, &episodes)
+fmt.Println(episodes)
 ```
 
-If you think back to the [first tutorial in the series](https://), the `Find` function might look similar to the `ListDatabaseNames` function. We can provide a context and some query parameters, and get our results. In this example `bson.M` represents a map of fields in no particular order. However, because we're trying to return all Documents, there aren't any fields in our query.
+If you think back to the [first tutorial in the series](https://www.mongodb.com/blog/post/quick-start-golang--mongodb--starting-and-setup), the `Find` function might look similar to the `ListDatabaseNames` function. We can provide a context and some query parameters, and get our results. In this example `bson.M` represents a map of fields in no particular order. However, because we're trying to return all documents, there aren't any fields in our query.
 
-Assuming no error happens, the results will exist in a MongoDB Cursor. In this simple example, the Cursor is iterated and each Document within the iteration is printed. Remember, for now at least, `bson.D` represents a MongoDB BSON Document.
+Assuming no error happens, the results will exist in a MongoDB cursor. In this simple example, all documents are decoded into a `[]bson.M` variable, the cursor is closed, and then the variable is printed.
 
 To get an idea of what the printed data would look like, take the following for example:
 
 ```
-[{_id ObjectID("5d9f4701e9948c0f65c9165d")} {podcast ObjectID("5d9e0173c1305d2a54eb431a")} {title GraphQL for API Development} {description Learn about GraphQL from the co-creator of GraphQL, Lee Byron.} {duration 25}]
-[{_id ObjectID("5dbb21faecb6c4837b0eff62")} {podcast ObjectID("5dbb2038e7841e4744f57a9c")} {title Progressive Web Application Development} {description Learn about PWA development with Tara Manicsic.} {duration 32}]
+[map[_id:ObjectID("5dc98dea14c897e9ab808161") description:This is the first episode. duration:25 podcast:ObjectID("5dc98de914c897e9ab808160") title:Episode #1] map[_id:ObjectID
+("5dc98dea14c897e9ab808162") description:This is the second episode. duration:32 podcast:ObjectID("5dc98de914c897e9ab808160") title:Episode #2]]
+```
+
+If your expected result set is large, using the `*mongo.Cursor.All` function might not be the best idea. Instead, you can iterate over your cursor and have it retrieve your data in batches. To do this, our code might change to the following:
+
+```go
+cursor, err := episodesCollection.Find(ctx, bson.M{})
+if err != nil {
+    log.Fatal(err)
+}
+for cursor.Next(ctx) {
+    var episode bson.M
+    if err = cursor.Decode(&episode); err != nil {
+        log.Fatal(err)
+    }
+    fmt.Println(episode)
+}
+cursor.Close(ctx)
```

Comment:
This `Close` call should be deferred so an error from `Decode` doesn't cause a leak.

Reply:
Added `defer` and moved above the loop.